engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSinkFactoryV2.scala

```diff
@@ -0,0 +1,152 @@
+package pl.touk.nussknacker.engine.avro.sink
+
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.{CustomNodeError, NodeId}
+import pl.touk.nussknacker.engine.api.context.transformation.{DefinedEagerParameter, NodeDependencyValue}
+import pl.touk.nussknacker.engine.api.context.{ProcessCompilationError, ValidationContext}
+import pl.touk.nussknacker.engine.api.definition._
+import pl.touk.nussknacker.engine.api.process.{ProcessObjectDependencies, SinkFactory}
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedObjectTypingResult, TypingResult}
+import pl.touk.nussknacker.engine.api.{LazyParameter, MetaData}
+import pl.touk.nussknacker.engine.avro.schemaregistry.SchemaRegistryProvider
+import pl.touk.nussknacker.engine.avro.{KafkaAvroBaseTransformer, SchemaDeterminerErrorHandler}
+import pl.touk.nussknacker.engine.avro.KafkaAvroBaseTransformer.{SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, SinkValueParamName, TopicParamName}
+import pl.touk.nussknacker.engine.avro.typed.AvroSchemaTypeDefinitionExtractor
+import pl.touk.nussknacker.engine.flink.api.process.FlinkSink
+import pl.touk.nussknacker.engine.api.typed
+import pl.touk.nussknacker.engine.definition.parameter.editor.ParameterTypeEditorDeterminer
+
+object KafkaAvroSinkFactoryV2 {
+
+  private val restrictedParamNames = Set(SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, TopicParamName)
+
+  private def containsRestrictedNames(obj: TypedObjectTypingResult): Boolean =
+    (obj.fields.keySet & restrictedParamNames).nonEmpty
+
+  private def toParamEither(typing: TypingResult)(implicit nodeId: NodeId): Validated[ProcessCompilationError, Either[List[Parameter], Parameter]] = {
```

Comment:
Why do we need this Either? Why Parameter is not just list with one element?

Reply:
We need a distinction between object with one field and single value. I've replaced Either with new abstraction